### PR TITLE
albef retrieval finetune

### PR DESCRIPTION
**Pull request by katrina433**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #229
* __->__ #228
* #227
* #226
* #225
* #224

Differential Revision: [D38881556](https://our.internmc.facebook.com/intern/diff/D38881556)

### CODE REVIEW

File: examples/albef/finetune_retrieval.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+        image_embed, image_feat = model(image=image, input_type="image", is_train=False)
+        image_embeds.append(image_embed)
+        image_feats.append(image_feat)
+    image_embeds = torch.cat(image_embeds, dim=0)
+    image_feats = torch.cat(image_feats, dim=0)
+
+    sims_matrix = image_feats @ text_feats.t()
+    image_to_text_scores = torch.full((num_images, num_text), -100.0).to(device)
+
+    num_tasks = dist_utils.get_world_size()
```
Comment (ebsmothers) [inline examples/albef/finetune_retrieval.py:138]: nit: can we just call this world_size or something?

Comment (ebsmothers) [inline examples/albef/finetune_retrieval.py:138]: nit: can we just call this world_size or something?

File: examples/albef/finetune_retrieval.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+            "epoch": epoch,
+        }
+        torch.save(
+            save_obj,
+            os.path.join(checkpoint_root, "retrieval_checkpoint_%02d.pth" % epoch),
+        )
+
+
+@torch.no_grad()
+def evaluation(model, datamodule, k, device):
```
Comment (ebsmothers) [inline examples/albef/finetune_retrieval.py:98]: Maybe split this up into multiple methods for handling text features, image features, and text_to_image/image_to_text_scores? As is it is a bit hard to follow

File: examples/albef/finetune_retrieval.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+    random.seed(seed)
+    cudnn.benchmark = True
+
+    datamodule = RetrievalDataModule(**config["datamodule_args"])
+    model = albef_model_for_retrieval(config)
+
+    pretrained_checkpoint_path = os.path.join(
+        config["checkpoint_root"], config["pretrained_checkpoint"]
+    )
+    model.load_state_dict(torch.load(pretrained_checkpoint_path, map_location="cpu"))
```
Comment (ebsmothers) [inline examples/albef/finetune_retrieval.py:279]: In the future you will have to move to torch.hub.load_state_dict_from_url once the checkpoint is on AWS

File: examples/albef/finetune_retrieval.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+        "img_r10": ir10,
+        "img_r_mean": ir_mean,
+        "r_mean": r_mean,
+    }
+    return eval_result
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--config", default="./examples/albef/configs/retrieval.yaml")
```
Comment (ebsmothers) [inline examples/albef/finetune_retrieval.py:254]: Will this default work if run from a different directory? If not, maybe you can use os.getcwd() or something similar

File: examples/albef/finetune_retrieval.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+    )
+    val_result = itm_eval(
+        image_to_text_scores,
+        text_to_image_scores,
+        datamodule.image_dataset.image_to_text,
+        datamodule.text_dataset.text_to_image,
+    )
+    result = {
+        "image_to_text_scores": image_to_text_scores,
+        "text_to_image_scores": text_to_image_scores,
```
Comment (ebsmothers) [inline examples/albef/finetune_retrieval.py:297]: Do you want to save all these? Seems it could be unnecessary/large

**Review by ebsmothers — APPROVED** [review]

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38881556).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38881556).

File: examples/albef/finetune_retrieval.py (hunk 4 above)
Comment (langong347) [reply examples/albef/finetune_retrieval.py:351]: do you want to correct this yourself? Just to flag this. Your call.

**Review by langong347 — APPROVED** [review]

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38881556).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38881556).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38881556).
